extendr-api/src/robj/mod.rs

```diff
@@ -221,7 +225,9 @@ impl Length for Robj {}
 impl Robj {
     pub fn from_sexp(sexp: SEXP) -> Self {
         single_threaded(|| {
-            unsafe { ownership::protect(sexp) };
+            if !PERMANENT.contains(&sexp) {
+                unsafe { ownership::protect(sexp) };
+            }
```

Comment:
This looks like a significant performance hit, performing a lookup every time we construct something from SEXP.
Alternatively, shouldn't this be part of `ownership::protect()`?

Reply:
It is still a performance hit if I put it in `ownership::protect`.
I don't know, I'm exploring options and trying to make my discoveries clear. 🤷

Reply:
I mean, this logic is part of the protection mechanism, not of the `Robj` creation.

Reply:
I agree with Ilia. cpp11 also does the `NULL`-check on the protection mechanism's side.

https://github.com/r-lib/cpp11/blob/500f642b4ea132ec8c168fc70a28e81e9510ece3/inst/include/cpp11/protect.hpp#L252-L254

Regarding the performance, I'm not confident, but I guess this won't matter much. The protection mechanism anyway need to check the entry in the `HashMap` if not rejecting `NULL`, and I don't think it's significantly slower or faster than comparing the pointer addresses. I might be wrong.

Reply:
In the limit, there is definitely a moment when comparing 4-5 pointers becomes as slow as looking up a value in a hash map of 4-5 elements, but for now we can ignore that.